### PR TITLE
Adding lidar support in the Bundle Adjustment

### LINKED ISSUE #4705: Update BundleAdjust and jigsaw for LiDAR
**Description**  

Update BundleAdjust and jigsaw with the changes from the LroLidar_Infrastructure branch. This should include the changes for the LiDAR work, but not the piecewise polynomial work. See PR #69, #75, #200, #280, #301, #3172 and #3250.

### PR DESCRIPTION
**Pull request by jessemapel**
## Description

Adds the ability to constrain observations with lidar range measurements in jigsaw. This only supports LOLA data via the lrolola2isis application. This work was original done by a handful of people: Ken Edmundson and Debbie Cook wrote the jigsaw/bundle adjust code; @scsides, Ian Humphrey, and Makayla Shepard wrote the lidar data format and lrolola2isis.

The original code was manually merged into dev as it contained other changes that we did not want to integrate.

## Related Issue

Fixes #4705

See the issue for the original PRs

## Motivation and Context

Range constraints help improve jigsaw solutions by ensuring that the solution better matches physical reality. This PR provides limited functionality, but does produce improved results for support data sets. It is also a good starting point for future lidar data sets and different types of constraints.

## How Has This Been Tested?

Extensive automated tests have been added. We also generated a LRO NAC stereo DEM with and without the lidar constraint and saw an improvement in the vertical alignment of the data.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Documentation change (update to the documentation; no code change)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] I have read and agree to abide by the [Code of Conduct](https://github.com/USGS-Astrogeology/ISIS3/blob/dev/Code-Of-Conduct.md)
- [x] I have read the [**CONTRIBUTING**](https://github.com/USGS-Astrogeology/ISIS3/blob/dev/CONTRIBUTING.md) document.
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.
- [x] I have added myself to the [.zenodo.json](https://github.com/USGS-Astrogeology/ISIS3/blob/dev/.zenodo.json) document.
- [x] I have added any user impacting changes to the [CHANGELOG.md](https://github.com/USGS-Astrogeology/ISIS3/blob/dev/CHANGELOG.md) document.

## Licensing
This project is mostly composed of free and unencumbered software released into the public domain, and we are unlikely to accept contributions that are not also released into the public domain. Somewhere near the top of each file should have these words:

> This work is free and unencumbered software released into the public domain. In jurisdictions that recognize copyright laws, the author or authors of this software dedicate any and all copyright interest in the software to the public domain.

- [x] I dedicate any and all copyright interest in this software to the public domain. I make this dedication for the benefit of the public at large and to the detriment of my heirs and successors. I intend this dedication to be an overt act of relinquishment in perpetuity of all present and future rights to this software under copyright law.

### CODE REVIEW

File: isis/src/control/objs/BundleAdjust/BundleAdjust.h [hunk 1]
```diff
@@ -387,7 +404,8 @@ namespace Isis {
       bool computeBundleStatistics();
       void applyParameterCorrections();
       bool errorPropagation();
-      double computeResiduals();
+      void computeResiduals();
```
Comment (AustinSanders) [inline isis/src/control/objs/BundleAdjust/BundleAdjust.h:407]: Changing return type constitutes API breaking change

File: isis/src/control/objs/BundleAdjust/BundleAdjust.h [hunk 2]
```diff
@@ -427,39 +448,32 @@ namespace Isis {
...
+      bool invert3x3(LinearAlgebra::MatrixUpperTriangular &m);
+      bool productATransB(LinearAlgebra::MatrixUpperTriangular &N22,
+                          SparseBlockColumnMatrix              &N12,
+                          SparseBlockRowMatrix                 &Q);
 
       // CHOLMOD library methods
 
       bool initializeCHOLMODLibraryVariables();
       bool freeCHOLMODLibraryVariables();
-      bool cholmodInverse();
```
Comment (AustinSanders) [inline isis/src/control/objs/BundleAdjust/BundleAdjust.h:445]: API breaking function removal

File: isis/src/control/objs/BundleAdjust/BundleAdjust.h [hunk 3]
```diff
@@ -427,39 +448,32 @@ namespace Isis {
...
       bool loadCholmodTriplet();
-      bool wrapUp();
```
Comment (AustinSanders) [inline isis/src/control/objs/BundleAdjust/BundleAdjust.h:447]: API breaking function removal

File: isis/src/control/objs/ControlPoint/ControlPoint.h [hunk 4]
```diff
@@ -591,19 +593,19 @@ namespace Isis {
...
+    protected:
+      void PointModified();
+      //!< List of Control Measures
+      QHash< QString, ControlMeasure * > * measures;
+
     private:
       void SetExplicitReference(ControlMeasure *measure);
       void ValidateMeasure(QString serialNumber) const;
       void AddMeasure(ControlMeasure *measure);
-      void PointModified();
```
Comment (AustinSanders) [inline isis/src/control/objs/ControlPoint/ControlPoint.h:598]: API breaking function removal

Comment (AustinSanders) [reply isis/src/control/objs/ControlPoint/ControlPoint.h:598]: Don't need to worry about private API

**Review by AustinSanders — COMMENTED** [review]
I gave this another quick look, and I'm not able to find any public API breaks

**Comment by jessemapel** [comment]
@AustinSanders Getting build errors in the tests. Can you check that it merged correctly and doesn't need any new headers or anything

**Comment by jessemapel** [comment]
@AustinSanders Tests on this are good now.

**Review by AustinSanders — APPROVED** [review]

**Comment by oleg-alexandrov** [comment]
Good work. What I was not able to see after inspecting the whole pull request (but it is big so I could have missed it), is a document spelling out what this tool does, how to use it, how the constraints are defined, how they are enforced, and ideally a solved example having all inputs and outputs.

**Comment by jessemapel** [comment]
There is documentation for lrolola2isis on how to use it to ingest the lidar csv files. Then, there's documentation on jigsaw for how to pass the lidar data in and set uncertainties.

I'm working on a PR to update the jigsaw documentation with a lot more detail and I can also add an example for using lidar when I do that.

**Comment by jessemapel** [comment]
Right now, I can just say that the constraints are added as additional observations with weights set by the user.

**Comment by oleg-alexandrov** [comment]
An example would be very nice, thank you, including how to use  lrolola2isis, how to call jigsaw, some other commands, if needed, and maybe even a pointer to a dataset that was shown to work. 

I assume then that the weights are used as multipliers to some additional terms to the cost function, so something like weight * some_error^2, rather than a hard range constraint, which would prevent a quantity from exiting an expected interval.

**Comment by jessemapel** [comment]
> I assume then that the weights are used as multipliers to some additional terms to the cost function, so something like weight * some_error^2, rather than a hard range constraint, which would prevent a quantity from exiting an expected interval.

Correct. They work just like all the other weights on the bundle.

**Pull request merged**